internal/backend/runtime/omni/audit/log_file.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package audit
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/siderolabs/gen/pair/ordered"
+)
+
+// LogFile is a rotating log file.
+//
+//nolint:govet
+type LogFile struct {
+	dir string
+
+	mu        sync.Mutex
+	f         *os.File
+	lastWrite time.Time
+}
+
+// NewLogFile creates a new rotating log file.
+func NewLogFile(dir string) *LogFile {
+	return &LogFile{dir: dir}
+}
+
+// Write writes data to the log file, creating new one on demand.
+func (l *LogFile) Write(data any) error {
+	return l.writeAt(data, time.Now())
+}
+
+func (l *LogFile) writeAt(data any, at time.Time) error {
+	f, err := l.openFile(at)
+	if err != nil {
+		return err
+	}
+
+	if err := json.NewEncoder(f).Encode(data); err != nil {
```

Comment:
this won't work well - we use direct file (without buffering), and json encoder would issue tons of small writes directly to the OS.

plus this won't be atomic, as append mode is only guaranteed for a single write (records will be interleaved)

in this case I would say it's easier to to write to the `bytes.Buffer` and then write bytes as a single write to the file.

Reply:
Yeah, I agree.